design-proposals/klip-11-DQL.md

```diff
@@ -1 +1,745 @@
+
+Note, that the actual terminology is base on the query **input and output type**! For example, the **table** query
+
+```sql
+SELECT * FROM myTable
+```
+
+has an input and output table. If the query is "converted" into a **stream** query via
+
+```sql
+SELECT * FROM STREAM(SELECT * FROM myTable)
+```
+
+we have two queries: and inner table query and an outer stream query, and the stream query has a stream as its input and output type.
+Hence, even if we allow a shorter notation as syntactic sugar:
+
+```sql
+STREAM(SELECT * FROM myTable)
+```
+
+it's still two nested queries and not one query with input table and output stream (one could think of the output type as the determining factor,
+if a query is called a stream or table query—however, that is somewhat a shortcut---not sure if this might lead to confusion?).
+
+
+
+#### Terminal Stream Queries:
+
+While a table query is always terminal, a stream query is continuous by default but may be terminal.
+The below examples illustrate how a custom start and end point for stream queries could be specified base on either time or offset:
+
+```sql
+-- setting custom starting point
+SELECT * FROM stream
+  WHERE ROWTIME >= _someTimestamp_
+        [AND|OR <some filter conditions over the attributes>];
+
+-- we need to discuss how we handle partitions; the idea atm is, to apply the same offset to all partitions
+-- (it's questionable how useful this is, but I think we should allow it anyway)
+SELECT * FROM stream
+  WHERE OFFSET >= _someNumber_
+        [AND|OR <some filter conditions over the attributes>];
+
+-- setting custom end point (this defines a terminal stream query)
+
+-- note, this will terminate when event time reaches current wall-clock time
+--   => this might not be intuitive, but also not too hard to explain to users
+SELECT * FROM stream
+  WHERE ROWTIME <= now()
+        [AND|OR <some filter conditions over the attributes>];
+
+-- terminates when partition end-offset is reached
+--   => is this query what some people mean by a "point in time over stream" query ???
+SELECT * FROM stream
+  WHERE OFFSET <= end()
+        [AND|OR <some filter conditions over the attributes>];
+
+
+-- join
+-- setting custom starting and end point is the same as above with corresponding `stream1.` or `stream2.` prefix
+SELECT *
+  FROM stream1 JOIN stream2
+    WITHIN _someTimeWindow_
+    ON [join-condition]
+  WHERE <some filter conditions over the attributes>];
+```
+
+Open questions:
+
+* Should we allow persistent, terminal, stream queries?
+  * In general, I think yes (already mentioned in the beginning)
+* It seems that `ROWTIME <= now()` semantics might not be intuitive as it mixes event-time and wall-clock time:
+  * I would expect that people often just want to stop at the end of the topic and confuse both concepts
+  * can we do better?
+* Is `OFFSET <= end()` too clumsy to express what many people may want, i.e., terminate at the end of the topic?
```

Comment:
I like that idea, even if the syntax is not great IMHO, as `<=` has no straight forward semantics for pairs. If `end()` returns  pair `(p,o)` we want hat `PARTITION = p AND OFFSET <= o` what is not really expressed above. (Also, users might want to use `<` and `<=` and both should be supported IMHO.

Maybe we can have `WHERE OFFSET <= endOfPartition()` or `WHERE OFFSET <= end(PARTITION)` ?

A more fundamental question that I also excluded in this KLIP is, if/how we want to expose metadata like OFFSET and PARTITION to begin with. Do we consider them as part of the _schema_? Or should we have some _functions_ that return the offset/partition? As offsets/partitions and immutable, they should also be handled differently than timestamps and headers IMHO -- at least for headers, we need to have a way for people to modify them (maybe even for timestamps). This might be a separate KLIP though...

Reply:
> Do we consider them as part of the schema

No. They should should be part of the schema.  Having them in the schema causes problems.  `SELECT * FROM FOO` should not include any metadata by default. If the user wants to access metadata columns then they should be doing:

```sql
SELECT *, offset() AS sourceOffset, partition() AS sourcePartition FROM FOO;
```

Reply:
> f the syntax is not great IMHO, as <= has no straight forward semantics for pairs. If end() returns pair (p,o) we want hat PARTITION = p AND OFFSET <= o what is not really expressed above

Agreed, syntax needs work: it needs to fit our type system and its semantics. However, I'm just trying to express the concept.

```sql
... WHERE partitionOffset() < latestPartitionOffset();
```

??

Reply:
The argument that `offset`, `partition`, and `topics` are not part of the schema makes sense to me.

It's still unclear to me, if we want to allow timestamp modification, ie, setting `ROWTIME` explicitly? What alternatives do we have to model `ROWTIME` -- the current implicit column is different to a _function_ that return the `timestamp()` -- why do we want/need this distinction?

Similar question about headers?